src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -91,34 +106,153 @@ def self.compile_ast(code, compile_options, boltlib_path)
+            pal.with_catalog_compiler(**variables) do |compiler|
+              # TODO: PUP-10476 Explore setting these as default in PAL. They are the defaults in Puppet
+              Puppet[:strict] = :warning
+              Puppet[:strict_variables] = false
 
+              ast = build_program(code)
               compiler.evaluate(ast)
+              compiler.evaluate_ast_node
               compiler.compile_additions
               compiler.catalog_data_hash
             end
           end
         end
       end
-      private_class_method :compile_ast
+      private_class_method :compile_for_environment_plan
+
+      # When we do not need to load bolt we assume there are no Bolt types to resolve in
+      # AST compilation.
+      def self.compile_for_environment(code, compile_options, boltlib_path)
+        # Set the node_name_value setting directly
+        Puppet[:node_name_value] = compile_options['certname']
+
+        # Use the existing environment with the requested name
+        Puppet::Pal.in_environment(compile_options['environment'],
+                                    envpath: Puppet[:environmentpath],
+                                    facts: compile_options['facts']['values'],
+                                    variables: compile_options['variables']['values']) do |pal|
+          Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
+
+          # This compiler has been configured with a node containing
+          # the requested environment, facts, and variables, and is used
+          # to compile a catalog in that context from the supplied AST.
+          pal.with_catalog_compiler do |compiler|
+            # We have to parse the AST inside the compiler block, because it
+            # initializes the necessary type loaders for us.
+            ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
+
+            compiler.evaluate(ast)
+            compiler.compile_additions
+            compiler.catalog_data_hash
+          end
+        end
+      end
+      private_class_method :compile_for_environment
+
+      # Prior to PE-29443 variables were in a hash. Serialization between ruby/clojure/json did
+      # not preserve hash order necessary for deserialization. The data strucutre is now stored
+      # in a list for moving data and the list is used to construct an ordered ruby hash.
+      def self.ordered_plan_vars(compile_options)
+        if compile_options['variables']['values'].is_a?(Array)
+          compile_options['variables']['values'].each_with_object({}) do |param_hash, acc|
+            acc[param_hash.keys.first] = param_hash.values.first
+          end
+        else
+          compile_options['variables']['values']
+        end
+      end
+      private_class_method :ordered_plan_vars
+
+      # Create a puppet settings context that can compile for projects
+      def self.with_puppet_settings_for_projects(overrides = {}, multithreaded)
```

Comment:
Instead of passing `multithreaded` all the way from the clojure config, we could we infer it from the class type of the `Puppet.settings` object instead. If `multithreaded` is on, then it will be a `Puppet::Server::Settings` class; otherwise, it will be a `Puppet::Settings` class.

You have already done the hard work to thread it through and it looks correct/working, but it might be simpler to just deduce it within the Ruby worker itself.

Reply:
Alternatively, you could just change this to a instance variable so it is available for AST compilation:
https://github.com/puppetlabs/puppetserver/blob/02ffd0ff877f5e83398dac9bb6c4d1c6827b1aa6/src/ruby/puppetserver-lib/puppet/server/master.rb#L37

Reply:
I thought about inferring multithreaded from the type, but I felt like that coupled together the multithreaded setting with the class definition and figured it would be at least a little better to keep those things independent where we can.

I do like the instance variable option though, I hadn't looked through the master class enough to realize I could just pass it in there.